File: backend/app/gateway/desktop_bridge.py
"""
DesktopBridge - WebSocket 代理让后端命令穿过前端到用户 macOS

架构：
  Backend Agent → WebSocket → Frontend → 用户 macOS 执行 → 结果返回

前端连接时携带认证信息（session_token），后端通过 WebSocket 发命令，
前端在用户本机执行后返回结果。

支持两类远程执行：
1. AppleScript / 系统命令（macOS）
2. CDP browser 操作（通过 browser.py 的 CDP 模式）
"""

import asyncio
import json
import logging
from typing import Optional, Dict, Any
from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from starlette.websockets import WebSocketState

logger = logging.getLogger(__name__)

router = APIRouter(tags=["desktop"])


class DesktopBridgeServer:
    """
    服务端：管理所有前端 WebSocket 连接，转发命令到对应连接。
    """

    def __init__(self):
        self._connections: Dict[str, WebSocket] = {}  # session_id -> websocket
        self._lock = asyncio.Lock()

    async def connect(self, session_id: str, ws: WebSocket):
        async with self._lock:
            old = self._connections.get(session_id)
            if old and old.client_state == WebSocketState.CONNECTED:
                try:
                    await old.close()
                except Exception:
                    pass
            await ws.accept()
            self._connections[session_id] = ws
            logger.info(f"[DesktopBridge] 前端连接: session_id={session_id}")

    async def disconnect(self, session_id: str):
        async with self._lock:
            self._connections.pop(session_id, None)
            logger.info(f"[DesktopBridge] 前端断开: session_id={session_id}")

    async def send_command(self, session_id: str, command: dict) -> Optional[dict]:
        """向指定 session 的前端发送命令，等待执行结果"""
        ws = self._connections.get(session_id)
        if not ws or ws.client_state != WebSocketState.CONNECTED:
            return {"status": "error", "message": "前端未连接"}

        try:
            await ws.send_json({"type": "desktop_cmd", **command})
        except Exception as e:
            logger.error(f"[DesktopBridge] 发送命令失败: {e}")
            return {"status": "error", "message": str(e)}

        # 等待结果（前端回复 type=desktop_result）
        try:
            # 超时 60s
            result = await asyncio.wait_for(ws.receive_json(), timeout=60)
            return result
        except asyncio.TimeoutError:
            return {"status": "error", "message": "执行超时（60s）"}
        except Exception as e:
            return {"status": "error", "message": str(e)}

    async def broadcast(self, message: dict):
        """广播消息到所有连接的前端（用于通知）"""
        async with self._lock:
            for ws in list(self._connections.values()):
                if ws.client_state == WebSocketState.CONNECTED:
                    try:
                        await ws.send_json(message)
                    except Exception:
                        pass


# 全局单例
_bridge = DesktopBridgeServer()


@router.websocket("/ws/desktop/{session_id}")
async def desktop_websocket(ws: WebSocket, session_id: str):
    """
    前端 WebSocket 连接端点。

    连接建立后，前端处于等待命令状态。
    后端通过 send_command() 向前端推送命令，前端执行后返回结果。

    消息格式：
      前端 → 后端：{"type": "desktop_result", "cmd_id": "...", "status": "success", "result": "..."}
      后端 → 前端：{"type": "desktop_cmd", "cmd_id": "...", "action": "osascript", "script": "..."}
    """
    await _bridge.connect(session_id, ws)
    try:
        while True:
            data = await ws.receive_json()
            msg_type = data.get("type")

            if msg_type == "desktop_result":
                # 前端主动上报结果（暂时不需要，前端回复 await send_command 即可）
                logger.debug(f"[DesktopBridge] 前端结果: {data}")

            elif msg_type == "ping":
                await ws.send_json({"type": "pong", "time": data.get("time")})

            else:
                logger.warning(f"[DesktopBridge] 未知消息类型: {msg_type}")

    except WebSocketDisconnect:
        logger.info(f"[DesktopBridge] WebSocket 断开: session_id={session_id}")
    except Exception as e:
        logger.error(f"[DesktopBridge] WebSocket 错误: {e}", exc_info=True)
    finally:
        await _bridge.disconnect(session_id)


@router.post("/api/desktop/execute")
async def desktop_execute(
    session_id: str,
    action: str,
    params: Optional[dict] = None,
):
    """
    HTTP 接口：后端 Agent 执行 desktop 命令。

    后端 Agent 调用此接口，将命令转发给已连接的前端代理执行。
    适用于 browser 有头模式（CDP）和 desktop 工具。

    Params:
      session_id: 前端会话 ID（用于定位 WebSocket 连接）
      action: desktop / browser_cdp
      params: 执行参数

    Returns:
      {"status": "success", "result": "..."} 或 {"status": "error", "message": "..."}
    """
    params = params or {}

    if action == "desktop":
        # desktop 工具走 AppleScript/syscommand
        from app.tools.implementations.desktop import desktop_execute as _desktop_exec
        try:
            result = await _desktop_exec(**params)
            return {"status": "success", "result": result}
        except Exception as e:
            return {"status": "error", "message": str(e)}

    elif action == "browser_cdp":
        # browser CDP 模式：复用 desktop WebSocket 代理 CDP 命令
        # params 应包含 cdp_url, action, 和 action 参数
        cdp_url = params.get("cdp_url")
        browser_action = params.get("browser_action", "navigate")
        browser_params = params.get("browser_params", {})

        if not cdp_url:
            return {"status": "error", "message": "browser_cdp 需要 cdp_url"}

        # 通过 WebSocket 让前端帮忙执行 CDP
        cmd = {
            "cmd_id": f"cdp_{session_id}_{browser_action}",
            "action": "browser_cdp",
            "cdp_url": cdp_url,
            "browser_action": browser_action,
            "browser_params": browser_params,
        }
        result = await _bridge.send_command(session_id, cmd)
        return result

    else:
        return {"status": "error", "message": f"未知 action: {action}"}

File: frontend/src/services/DesktopBridge.ts
/**
 * DesktopBridge 前端服务
 *
 * 前端作为 WebSocket 代理：
 * 1. 建立 WebSocket 连接到后端 /ws/desktop/{session_id}
 * 2. 接收后端发来的 desktop_cmd 命令
 * 3. 在用户本机执行（AppleScript / 系统命令 / CDP）
 * 4. 将结果返回给后端
 */

export interface DesktopCommand {
  cmd_id: string;
  action: 'osascript' | 'bash' | 'cliclick' | 'browser_cdp' | 'screencapture';
  script?: string;       // osascript 时用
  command?: string;      // bash 时用
  args?: string[];       // cliclick 时用
  cdp_url?: string;      // browser_cdp 时用
  browser_action?: string;
  browser_params?: Record<string, any>;
  path?: string;         // screencapture 路径
}

export interface DesktopResult {
  cmd_id: string;
  status: 'success' | 'error';
  result?: string;
  error?: string;
}

type MessageHandler = (cmd: DesktopCommand) => void;

class DesktopBridgeService {
  private ws: WebSocket | null = null;
  private sessionId: string = '';
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pendingCommands = new Map<string, {
    resolve: (result: DesktopResult) => void;
    reject: (error: Error) => void;
    timeout: ReturnType<typeof setTimeout>;
  }>();
  private handlers: Set<MessageHandler> = new Set();
  private isConnected = false;

  /**
   * 连接后端 DesktopBridge WebSocket
   */
  connect(sessionId: string): Promise<void> {
    this.sessionId = sessionId;

    return new Promise((resolve, reject) => {
      const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
      const wsUrl = `${protocol}//${window.location.host}/ws/desktop/${sessionId}`;

      try {
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
          console.log('[DesktopBridge] WebSocket 已连接');
          this.isConnected = true;
          this.clearReconnectTimer();
          resolve();
        };

        this.ws.onmessage = async (event) => {
          try {
            const data = JSON.parse(event.data);
            await this.handleMessage(data);
          } catch (e) {
            console.error('[DesktopBridge] 消息解析失败:', e);
          }
        };

        this.ws.onerror = (error) => {
          console.error('[DesktopBridge] WebSocket 错误:', error);
          this.isConnected = false;
        };

        this.ws.onclose = () => {
          console.log('[DesktopBridge] WebSocket 断开');
          this.isConnected = false;
          this.scheduleReconnect();
        };
      } catch (err) {
        reject(err);
      }
    });
  }

  /**
   * 处理从后端收到的消息
   */
  private async handleMessage(data: any): Promise<void> {
    if (data.type === 'desktop_cmd') {
      const cmd: DesktopCommand = data as DesktopCommand;
      console.log('[DesktopBridge] 收到命令:', cmd);

      // 先检查是否有待处理的 pending 命令
      const pending = this.pendingCommands.get(cmd.cmd_id);

      try {
        const result = await this.executeCommand(cmd);

        if (pending) {
          pending.resolve({ cmd_id: cmd.cmd_id, status: 'success', result });
          this.pendingCommands.delete(cmd.cmd_id);
          clearTimeout(pending.timeout);
        } else {
          // 没有 pending，通过 WebSocket 发送结果
          await this.sendResult({ cmd_id: cmd.cmd_id, status: 'success', result });
        }
      } catch (err: any) {
        const errorMsg = err?.message || String(err);
        console.error('[DesktopBridge] 命令执行失败:', errorMsg);

        if (pending) {
          pending.resolve({ cmd_id: cmd.cmd_id, status: 'error', error: errorMsg });
          this.pendingCommands.delete(cmd.cmd_id);
          clearTimeout(pending.timeout);
        } else {
          await this.sendResult({ cmd_id: cmd.cmd_id, status: 'error', error: errorMsg });
        }
      }
    } else if (data.type === 'ping') {
      this.ws?.send(JSON.stringify({ type: 'pong', time: data.time }));
    }
  }

  /**
   * 执行 desktop 命令
   */
  private async executeCommand(cmd: DesktopCommand): Promise<string> {
    switch (cmd.action) {
      case 'osascript':
        return this.executeOsascript(cmd.script!);

      case 'bash':
        return this.executeBash(cmd.command!);

      case 'cliclick':
        return this.executeCliclick(cmd.args || []);

      case 'browser_cdp':
        return this.executeBrowserCDP(cmd);

      case 'screencapture':
        return this.executeScreencapture(cmd.path!);

      default:
        throw new Error(`未知命令类型: ${cmd.action}`);
    }
  }

  /**
   * 执行 AppleScript
   */
  private executeOsascript(_script: string): Promise<string> {
    return new Promise((_resolve, reject) => {
      reject(new Error('osascript 需要本地代理程序，请安装 DesktopBridge agent'));
    });
  }

  /**
   * 执行 bash 命令
   */
  private executeBash(_command: string): Promise<string> {
    return new Promise((_resolve, reject) => {
      reject(new Error('bash 需要本地代理程序，请安装 DesktopBridge agent'));
    });
  }

  /**
   * 执行 cliclick（鼠标控制）
   */
  private executeCliclick(_args: string[]): Promise<string> {
    return new Promise((_resolve, reject) => {
      reject(new Error('cliclick 需要本地代理程序，请安装 DesktopBridge agent'));
    });
  }

  /**
   * 通过 CDP 控制浏览器
   */
  private async executeBrowserCDP(cmd: DesktopCommand): Promise<string> {
    const { cdp_url, browser_action, browser_params } = cmd;

    if (!cdp_url) {
      throw new Error('browser_cdp 需要 cdp_url');
    }

    try {
      const response = await fetch(cdp_url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          id: Date.now(),
          method: this.mapBrowserAction(browser_action || ''),
          params: browser_params,
        }),
      });

      const result = await response.json();
      return JSON.stringify(result);
    } catch (err: any) {
      throw new Error(`CDP 执行失败: ${err.message}`);
    }
  }

  /**
   * 映射 browser_action 到 CDP method
   */
  private mapBrowserAction(action: string): string {
    const map: Record<string, string> = {
      'navigate': 'Page.navigate',
      'click': 'Input.dispatchMouseEvent',
      'type': 'Input.dispatchKeyEvent',
      'screenshot': 'Page.captureScreenshot',
      'get_text': 'Runtime.evaluate',
      'scroll': 'Input.dispatchMouseEvent',
    };
    return map[action] || action;
  }

  /**
   * 执行截图
   */
  private executeScreencapture(_path: string): Promise<string> {
    return new Promise((_resolve, reject) => {
      reject(new Error('screencapture 需要本地代理程序，请安装 DesktopBridge agent'));
    });
  }

  /**
   * 发送执行结果到后端
   */
  private async sendResult(result: DesktopResult): Promise<void> {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ type: 'desktop_result', ...result }));
    }
  }

  /**
   * 断开连接
   */
  disconnect(): void {
    this.clearReconnectTimer();
    this.pendingCommands.forEach((p) => {
      clearTimeout(p.timeout);
      p.reject(new Error('连接已断开'));
    });
    this.pendingCommands.clear();
    this.ws?.close();
    this.ws = null;
    this.isConnected = false;
  }

  /**
   * 注册命令处理器（用于主动推送模式）
   */
  onCommand(handler: MessageHandler): () => void {
    this.handlers.add(handler);
    return () => this.handlers.delete(handler);
  }

  private scheduleReconnect(): void {
    this.clearReconnectTimer();
    this.reconnectTimer = setTimeout(() => {
      if (!this.isConnected && this.sessionId) {
        console.log('[DesktopBridge] 尝试重连...');
        this.connect(this.sessionId).catch(() => {});
      }
    }, 5000);
  }

  private clearReconnectTimer(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }

  get connected(): boolean {
    return this.isConnected;
  }
}

// 单例导出
export const desktopBridge = new DesktopBridgeService();